api/cpp/include/slint.h

```diff
@@ -1049,6 +1049,182 @@ class MapModel : public Model<MappedModelData>
+                                     return sort_fn(*sorted_elem, changed_value);
+                                 });
+
+        insertion_point = sorted_rows.insert(insertion_point, changed_row);
+        auto inserted_row = std::distance(sorted_rows.begin(), insertion_point);
+
+        if (inserted_row == removed_row) {
+            target_model.row_changed(removed_row);
+        } else {
+            target_model.row_removed(removed_row, 1);
+            target_model.row_added(inserted_row, 1);
+        }
+    }
+    void row_removed(int first_removed_row, int count) override
+    {
+        if (sorted_rows_dirty) {
+            reset();
+            return;
+        }
+
+        std::vector<int> removed_rows;
+        removed_rows.reserve(count);
+
+        for (auto it = sorted_rows.begin(); it != sorted_rows.end();) {
+            if (*it >= first_removed_row) {
+                if (*it < first_removed_row + count) {
+                    removed_rows.push_back(std::distance(sorted_rows.begin(), it));
+                    it = sorted_rows.erase(it);
+                    continue;
+                } else {
+                    *it -= count;
+                }
+            }
+            ++it;
+        }
+
+        for (int removed_row : removed_rows) {
+            target_model.row_removed(removed_row, 1);
+        }
+    }
+    void reset() override
+    {
+        sorted_rows_dirty = true;
+        target_model.reset();
+    }
+
+    void ensure_sorted()
+    {
+        if (!sorted_rows_dirty) {
+            return;
+        }
+
+        sorted_rows.resize(source_model->row_count());
+        for (size_t i = 0; i < sorted_rows.size(); ++i)
+            sorted_rows[i] = i;
+
+        std::sort(sorted_rows.begin(), sorted_rows.end(), [this](int lhs_index, int rhs_index) {
+            auto lhs_elem = source_model->row_data(lhs_index);
+            auto rhs_elem = source_model->row_data(rhs_index);
+            return sort_fn(*lhs_elem, *rhs_elem);
+        });
+
+        sorted_rows_dirty = false;
+    }
+
+    std::shared_ptr<slint::Model<ModelData>> source_model;
+    std::function<bool(const ModelData &, const ModelData &)> sort_fn;
+    slint::SortModel<ModelData> &target_model;
+    std::vector<int> sorted_rows;
+    bool sorted_rows_dirty = true;
+};
+}
+
+/// The SortModel acts as an adapter model for a given source model by sorting all rows
+/// with by order provided by the given sorting function. The sorting function is called for
+/// pairs of elements of the source model.
+template<typename ModelData>
+class SortModel : public Model<ModelData>
+{
+    friend struct private_api::SortModelInner<ModelData>;
+
+public:
+    /// Constructs a new SortModel that provides a sorted view on the \a source_model by applying
+    /// the order given by the specified \a sort_fn.
+    SortModel(std::shared_ptr<Model<ModelData>> source_model,
+              std::function<bool(const ModelData &, const ModelData &)> sort_fn)
```

Comment:
(Also valid for the other models: We could consider having the function has a template parameter to the SortModel, then we wouldn't have the std::function allocation when there is a capture.   But probably it's fine like that because these function usually do not have a capture, and there are already lots of allocation otherwise)

Reply:
I was wondering about the same. If it becomes a template parameter of the SortModel, the *use* site isn't a simple `std::make_shared<slint::SortModel<int>>(...)` anymore. I think it would have to be a factory function instead of the constructor. That's what turned me off from pursuing this further.